### PR TITLE
Feature/role permissions expires

### PR DESCRIPTION
**Pull request by eramitgupta**
## Optional Expiration for Roles and Permissions

If you want to add something to it, please let me know
or if there is any bug 

This Laravel package introduces an optional expiration feature for roles and permissions, allowing developers to assign roles and permissions to users with a specified expiration date. This functionality provides greater control over user access, enabling you to manage permissions dynamically based on your application's needs.

### Features

- **Optional Expiration Dates**: When assigning roles or permissions, you can optionally specify an expiration date. If no expiration date is provided, the role or permission remains active indefinitely.
  
- **Flexible Management**: Users can have multiple roles and permissions, each with its own expiration date. This allows for granular control over user access based on changing requirements or time-limited privileges.
  
- **Automatic Expiration Handling**: The system automatically checks if a role or permission has expired. If a role or permission is expired, it will be considered inactive, preventing access to protected resources.

### Usage

To assign a role or permission with an expiration date, you can use the following methods:

```bash
php artisan erag:new-update  
```
Refer to the `RolePermissionSeeder.php.stub` for examples.

### Setting Expiration

If you don’t need an expiration, simply store `null` in the database:

```php
/**
 * Use as needed.
 * 
 * Example usages:
 * - now()->subDays(2) // Subtract 2 days from the current date
 * - now()->addDays(1) // Add 1 day to the current date
 * - now()->addHours(2) // Add 2 hours to the current date
 * - now()->addMinutes(5) // Add 5 minutes to the current date
 */
```

To check if a user has a valid role or permission, the system will automatically verify if the current date is less than the expiration date:

```php
if (hasRole('admin')) {
    dd('You are allowed to access');
} else {
    dd('You are not allowed to access');
}
  
if (auth()->user()->can('create-post')) {
    dd('You are allowed to access');
} else {
    dd('You are not allowed to access');
}
```

### Can be used for many permissions

```php
if (hasPermissions('post-create')) {
    dd('You are allowed to access');
} else {
    dd('You are not allowed to access');
}

if (hasPermissions('post-create|post-edit')) {
    dd('You are allowed to access');
} else {
    dd('You are not allowed to access');
}

if (hasPermissions('post-create,post-edit')) {
    dd('You are allowed to access');
} else {
    dd('You are not allowed to access');
}


### Benefits

- **Dynamic Access Control**: Easily manage user access based on time-limited roles and permissions.
  
- **Improved Security**: Reduce the risk of unauthorized access by setting expiration dates for roles and permissions.

- **Simplified Role Management**: Quickly adjust user permissions without the need for manual intervention when roles or permissions expire.

### Conclusion

The optional expiration feature for roles and permissions enhances your application's security and flexibility, making it easier to manage user access based on evolving requirements. Use this feature to ensure that users only have access to the resources they need for the time they need them.

### CODE REVIEW

**Review by herovish143 — CHANGES_REQUESTED** [review]

File: src/database/Seeder/RolePermissionSeeder.php.stub [hunk 1]
```diff
@@ -70,7 +84,7 @@ class RolePermissionSeeder extends Seeder
                 'name' => 'Admin',
                 'email' => 'admin@gmail.com',
                 'password' => Hash::make('admin'),
-                'roles' => ['admin'],
+                'roles' => ['admin', 'editor'],
```
[the remark this answers is not in the document]
Comment (eramitgupta) [reply src/database/Seeder/RolePermissionSeeder.php.stub]: I fixed conflicts

File: src/Traits/HasPermissionsTrait.php [hunk 2]
```diff
@@ -73,6 +105,8 @@ protected function hasPermission($permission): bool
 
     protected function getAllPermissions(array $permissions)
     {
-        return Permission::whereIn('name', $permissions)->with('roles')->get();
+        $permissionNames = collect($permissions)->map(fn ($permission) => is_object($permission) ? $permission->name : $permission);
```
[the remark this answers is not in the document]
Comment (eramitgupta) [reply src/Traits/HasPermissionsTrait.php:108]: I have updated
You made an excellent suggestion. Thank you!

File: src/Traits/HasPermissionsTrait.php [hunk 3]
```diff
@@ -73,6 +105,8 @@ protected function hasPermission($permission): bool
...
+
+        return Permission::whereIn('name', $permissionNames)->get();
```
[the remark this answers is not in the document]
Comment (eramitgupta) [reply src/Traits/HasPermissionsTrait.php:110]: opppps 🫨 
I made a mistake

File: src/PermissionServiceProvider.php [hunk 4]
```diff
@@ -22,6 +23,7 @@ public function register(): void
     {
         $this->commands([
             PublishPermissionMigrations::class,
+            NewUpdate::class,
```
[the remark this answers is not in the document]
Comment (Houbsi) [reply src/PermissionServiceProvider.php:27]: What happens when you change the migrations in a later stage again? PublishNewNewMigrations?

I would change the logic in PublishPermissionMigrations::class to check before publishing.

File: src/database/Seeder/RolePermissionSeeder.php.stub [hunk 5]
```diff
@@ -32,10 +33,18 @@ class RolePermissionSeeder extends Seeder
         $permissions = [
             'create-post',
```
[the remark this answers is not in the document]
Comment (eramitgupta) [reply src/database/Seeder/RolePermissionSeeder.php.stub]: I fixed conflicts, please check it once

**Pull request closed without merging**